[New feature]Introduce iOS multi-touch drag behavior

xu-baolin opened the pull request:
Fixes #38926

This patch implements the iOS behavior pointed out by @dkwingsmt at #38926 , which is also consistent with the performance of my settings application on the iPhone.

### iOS behavior (horizontal or vertical drag)

## Algorithm
When dragging: delta(combined) = max(i of n that are positive) delta(i) - max(i of n that are negative) delta(i)
It means that, if two fingers are moving +50 and +10 respectively, it will move +50; if they're moving at +50 and -10 respectively, it will move +40.

~~TODO~~
~~Write some test cases~~

xu-baolin committed:
Introduce iOS multi-touch drag behavior

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -385,12 +391,170 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
+      return;
+    }
+
+    final Duration currentSystemFrameTimeStamp = SchedulerBinding.instance.currentSystemFrameTimeStamp;
```
Fortunately, we can use this to identify whether the events are of the same batch. CC @dkwingsmt

dkwingsmt replied:
Brilliant!

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
add some tests

xu-baolin committed:
optimzation algorithm

Piinks commented on packages/flutter/lib/src/widgets/scroll_configuration.dart:
```diff
@@ -110,8 +110,20 @@ class ScrollBehavior {
-  MultitouchDragStrategy get multitouchDragStrategy => MultitouchDragStrategy.latestPointer;
+  /// create drag gestures for non-Apple platforms, and
+  /// [MultitouchDragStrategy.maxAllPointers] for Apple platforms.
+  MultitouchDragStrategy get multitouchDragStrategy {
```
I wonder if this should not be converted to a method so that is can have a BuildContext?

The defaultTargetPlatform is used by ScrollBehavior, but subclasses like MaterialScrollBehavior and CupertinoScrollBehavior use the getPlatform method using context to account for if the Theme is specifying a platform.

xu-baolin replied:
Good point.
I'm worried that this will become a breaking change unless we can guarantee that this will be released at the same time as the previous PR. What do you think?

Piinks replied:
They won't be released at the same time. The other PR has already been cut into multiple beta releases I think. 
It feels like a worthwhile breaking change though with low risk. Not a lot of folks will have had the opportunity to use the new getter, so following up with a small change to make it a method should be ok. I na help with a migration guide.

Piinks replied:
I think dart fix will be able to help too.

xu-baolin committed:
review feadback

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -495,6 +688,11 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
+
+    if (_activePointer == pointer) {
+      _activePointer =
+        _acceptedActivePointers.isNotEmpty ? _acceptedActivePointers.first : null;
```
Refer to Android [RecyclerView](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-master-dev/recyclerview/recyclerview/src/main/java/androidx/recyclerview/widget/RecyclerView.java)

![image](https://github.com/flutter/flutter/assets/61075224/34381d0c-02c6-43c1-8b9d-b4f2a2a101e4)

dkwingsmt commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -51,24 +51,42 @@ enum DragStartBehavior {
+  /// All active pointers will be tracked together. The scrolling offset
+  /// is determined by the maximum deltas of both directions.
+  ///
+  /// If drag with two pointers at the same time, with the first moving +10 pixels
+  /// and the second moving +20 pixels, the recognizer will report a delta of 20.
+  ///
+  /// If the first moving +20 pixels and the second moving -15 pixels, the
+  /// recognizer will report a delta of 5.
```
```suggestion
  /// If the user is dragging with 3 pointers at the same time, each having [+10, +20, +33] pixels of offset, the recognizer will report a delta of 33 pixels.
  ///
  /// If the user is dragging with 5 pointers at the same time, each having [+10, +20, +33, -1, -12] pixels of offset, the recognizer will report a delta of (+33) + (-12) = 21 pixels.
```

dkwingsmt commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -51,24 +51,42 @@ enum DragStartBehavior {
   /// speed will double or triple, depending on how many fingers are dragging
   /// at the same time.
+  ///
+  /// If drag with two pointers at the same time, with the first moving +20 pixels
+  /// and the second moving +20 pixels, the recognizer will report a delta of 40.
+  ///
+  /// If the first moving +20 pixels and the second moving -15 pixels, the
+  /// recognizer will report a delta of 5.
```
```suggestion
  /// If the user is dragging with 3 pointers at the same time, each having [+10, +20, +33] pixels of offset, the recognizer will report a delta of 10 + 20 + 33 = 63 pixels.
  ///
  /// If the user is dragging with 5 pointers at the same time, each having [+10, +20, +33, -1, -12] pixels of offset, the recognizer will report a delta of 10 + 20 + 33 - 1 - 12 = 50 pixels.
```

dkwingsmt commented on packages/flutter/lib/src/widgets/scroll_configuration.dart:
```diff
@@ -110,8 +110,20 @@ class ScrollBehavior {
+      case TargetPlatform.macOS:
+      case TargetPlatform.iOS:
+        return MultitouchDragStrategy.maxAllPointers;
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
```
Maybe?
```suggestion
      default:
```

dkwingsmt replied:
Is it idiomatic to have a default or to just enumerate all the cases and let the compiler check?

xu-baolin replied:
Consistent with elsewhere, an compilation error is thrown if forget to specify behavior for a new platform.

dkwingsmt reviewed (commented):
I haven't looked closely into the logic yet, just touching on the comments part.

dkwingsmt commented:
If I remember correctly, you said that you used the event timestamp to determine whether the current event belongs to the batch of the previous event. Can you briefly explain how this is done? Specifically, I asked this because we wanted to do the similar thing (but in the pointer event converter) and was concerned that, this approach requires that any calculation on a batch can not be started until the next pointer.

```
   event1    event2    event3    event4
   batch1    batch1    batch1    batch2
                                  ^
    wait...   wait...   wait...   | Hey a new timestamp! Time to process batch 1
```

This not introduces delay but also prevents the last batch from taking effect.

I assume your approach processes each event as soon as they arrive, and calculate incrementally based on whatever already is sent out on this batch?

xu-baolin commented:
1

发件人: Tong Mu ***@***.***>
发送时间: 2024年1月24日 4:39
收件人: flutter/flutter ***@***.***>
抄送: 徐宝林(Ava) ***@***.***>; Author ***@***.***>
主题: Re: [flutter/flutter] [New feature]Introduce iOS multi-touch drag behavior (PR #141355)


If I remember correctly, you said that you used the event timestamp to determine whether the current event belongs to the batch of the previous event. Can you briefly explain how this is done? Specifically, I asked this because we wanted to do the similar thing (but in the pointer event converter) and was concerned that, this approach requires that any calculation on a batch can not be started until the next pointer.

   event1    event2    event3    event4

   batch1    batch1    batch1    batch2

                                  ^

    wait...   wait...   wait...   | Hey a new timestamp! Time to process batch 1

This not introduces delay but also prevents the last batch from taking effect.

I assume your approach processes each event as soon as they arrive, and calculate incrementally based on whatever already is sent out on this batch?

—
Reply to this email directly, view it on GitHub<https://github.com/flutter/flutter/pull/141355#issuecomment-1906880363>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOR66GHPQHMF5BWSCIN3HXDYQANWLAVCNFSM6AAAAABBWAI6BWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWHA4DAMZWGM>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
OPPO

本电子邮件及其附件含有OPPO公司的保密信息，仅限于邮件指明的收件人（包含个人及群组）使用。禁止任何人在未经授权的情况下以任何形式使用。如果您错收了本邮件，切勿传播、分发、复制、印刷或使用本邮件之任何部分或其所载之任何内容，并请立即以电子邮件通知发件人并删除本邮件及其附件。
网络通讯固有缺陷可能导致邮件被截留、修改、丢失、破坏或包含计算机病毒等不安全情况，OPPO对此类错误或遗漏而引致之任何损失概不承担责任并保留与本邮件相关之一切权利。
除非明确说明，本邮件及其附件无意作为在任何国家或地区之要约、招揽或承诺，亦无意作为任何交易或合同之正式确认。 发件人、其所属机构或所属机构之关联机构或任何上述机构之股东、董事、高级管理人员、员工或其他任何人（以下称“发件人”或“OPPO”）不因本邮件之误送而放弃其所享之任何权利，亦不对因故意或过失使用该等信息而引发或可能引发的损失承担任何责任。
文化差异披露：因全球文化差异影响，单纯以YES\OK或其他简单词汇的回复并不构成发件人对任何交易或合同之正式确认或接受，请与发件人再次确认以获得明确书面意见。发件人不对任何受文化差异影响而导致故意或错误使用该等信息所造成的任何直接或间接损害承担责任。
This e-mail and its attachments contain confidential information from OPPO, which is intended only for the person or entity whose address is listed above. Any use of the information contained herein in any way (including, but not limited to, total or partial disclosure, reproduction, or dissemination) by persons other than the intended recipient(s) is prohibited. If you are not the intended recipient, please do not read, copy, distribute, or use this information. If you have received this transmission in error, please notify the sender immediately by reply e-mail and then delete this message.
Electronic communications may contain computer viruses or other defects inherently, may not be accurately and/or timely transmitted to other systems, or may be intercepted, modified ,delayed, deleted or interfered. OPPO shall not be liable for any damages that arise or may arise from such matter and reserves all rights in connection with the email.
Unless expressly stated, this e-mail and its attachments are provided without any warranty, acceptance or promise of any kind in any country or region, nor constitute a formal confirmation or acceptance of any transaction or contract. The sender, together with its affiliates or any shareholder, director, officer, employee or any other person of any such institution (hereinafter referred to as "sender" or "OPPO") does not waive any rights and shall not be liable for any damages that arise or may arise from the intentional or negligent use of such information.
Cultural Differences Disclosure: Due to global cultural differences, any reply with only YES\OK or other simple words does not constitute any confirmation or acceptance of any transaction or contract, please confirm with the sender again to ensure clear opinion in written form. The sender shall not be responsible for any direct or indirect damages resulting from the intentional or misuse of such information.

xu-baolin commented:
> I assume your algorithm takes a different approach and processes each event as soon as they arrive, and calculate incrementally based on whatever already is sent out on this batch?


@dkwingsmt, you are right.
In fact, all events before the next Vsync signal arrives are treated as a batch here. Each event will be processed immediately, without delay, and does not depend on the arrival of the last event. In other words, any event may be the last one in the batch.  Since there's no real start frame yet, we have the opportunity to merge handling of the input events.
![Image_20240124103122](https://github.com/flutter/flutter/assets/61075224/ca289094-c0bc-42e8-b8f6-cf912bab97b5)
![Image_20240124103132](https://github.com/flutter/flutter/assets/61075224/d799eaa2-c2fc-43dd-be4a-5cf2457f865b)

xu-baolin committed:
code review feedback

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -327,6 +328,10 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
 
   final Map<int, VelocityTracker> _velocityTrackers = <int, VelocityTracker>{};
 
+  // save the move deltas before persistentCallbacks
```
Change to:
`  // The move delta of each pointer before the next frame.`

xu-baolin committed:
fix the docs test

dkwingsmt reviewed (commented):
I read the algorithm in detail and gave some comments. Most of them are "this logic can be simplified" but I raised a major question in L448.

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -384,13 +395,192 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
+    }
+
+    for (final Offset delta in _moveDeltaBeforeFrame[pointer]!) {
+      if (positive) {
```
In fact, we need the sum delta of the positive and negative on both X and Y axes, Record the original data directly or customize the data structure to store the merged data. I think they are the same?

dkwingsmt replied:
I re-read the algorithm that I described and I removed the last paragraph.

But I still think we don't need to store original pointer events but only sum offset for each pointer. If a pointer moves slightly north then slightly south within one frame, why should half of the events be used to determine the positive range, while the other half used to determine the other half? I think the iOS logic is simply "get the northward-most pointer and the southward-most pointer".

xu-baolin replied:
I understand what you mean, let me try it.

xu-baolin replied:
What do you think the result of the following events should be?
pointer1:  (0, 15) + (0, -5)
pointer2:  (0, 15) + (0, 0)

xu-baolin replied:
I followed your suggestion and cache only one offset for each finger. The same finger moving back and forth within one frame seems less common and less important to the scrolling behavior.

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -384,13 +395,192 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
+      return localDelta;
+    }
+
+    final Duration currentSystemFrameTimeStamp = SchedulerBinding.instance.currentSystemFrameTimeStamp;
```
Good idea, I'll try it

xu-baolin replied:
![image](https://github.com/flutter/flutter/assets/61075224/e12c6212-fbf9-4aef-94db-73a51e404d72)
The event timestamp may changes during a frame.

`What happens if the screen is completely static (but there's still a gesture detector working, probably because the gesture's slop isn't reached yet).`

If the gesture is not accepted, the resolved delta will not applied.

xu-baolin committed:
review feedback

xu-baolin commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -384,13 +395,192 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
+
+  Offset _resolveLocalDeltaForMultitouch(int pointer, Offset localDelta) {
+    if (multitouchDragStrategy != MultitouchDragStrategy.maxAllPointers) {
+      _resetMoveDeltaRecords();
```
I reserved  `if (_frameTimeStamp != null) {` to prevent `clear()` from being called frequently.

dkwingsmt commented on packages/flutter/test/gestures/drag_test.dart:
```diff
@@ -648,6 +649,342 @@ void main() {
+    ]);
+  });
+
+  testGesture('Pan drag with multiple pointers - maxAllPointers', (GestureTester tester) {
```
I think this test case tested too much for the case that one pointer moves back and forth within a frame (we don't have to remove them, up to you), but doesn't test the case with more than 2 pointers. We want to test that the pointers that move not as much do not affect the resulting delta, and that they might become the border pointers when they move a lot.

dkwingsmt reviewed (commented):
Mostly ok, but another test case should be added.

xu-baolin committed:
review feedback 2

dkwingsmt commented on packages/flutter/test/gestures/drag_test.dart:
```diff
@@ -553,9 +554,9 @@ void main() {
   testGesture('Drag with multiple pointers in down behavior - default', (GestureTester tester) {
     final HorizontalDragGestureRecognizer drag1 =
-    HorizontalDragGestureRecognizer() ..dragStartBehavior = DragStartBehavior.down;
+      HorizontalDragGestureRecognizer() ..dragStartBehavior = DragStartBehavior.down;
```
What's with the indentations here and below?

xu-baolin replied:
Shouldn't it be indented here?

dkwingsmt commented on packages/flutter/test/gestures/drag_test.dart:
```diff
@@ -648,6 +649,601 @@ void main() {
+    tester.route(down6);
+
+    log.add('-b');
+    // #6 pointer move (110.0, 110.0), received delta should be (10, 10).
```
This is not the same as the panning algorithm I described in https://github.com/flutter/flutter/issues/38926, although I might have been wrong. Can you verify the native behavior of panning on iOS and see if it only checks the border pointers or if it averages all pointers? Also, what if the border pointers for x axis are different from those for y axis, for example:
```
5: Offset(-100, 0),
6: Offset(100, 0),
7: Offset(0, -100),
8: Offset(0, 100),
```
what's the behavior on native for cases like this?

xu-baolin replied:
I test the `UIPanGestureRecognizer` for iOS and indeed like you said, it takes the average of all fingers.
Here is the test code,
```dart
//
//  ViewController.swift
//  panning
//
//  Created by 徐宝林(Ava) on 2024/2/18.
//

import UIKit

class ViewController: UIViewController {
    
    private let pannableView: UIView = {
        // Initialize View
        let view = UIView(frame: CGRect(origin: .zero,
                                        size: CGSize(width: 300.0, height: 300.0)))

        // Configure View
        view.backgroundColor = .blue
        view.translatesAutoresizingMaskIntoConstraints = false

        return view
    }()
    
    @objc private func didPan(_ sender: UIPanGestureRecognizer) {
//        pannableView.center = sender.location(in: view)
        let center = pannableView.center
        let trans = sender.translation(in: view)
        pannableView.center = CGPointMake(center.x + trans.x, center.y + trans.y)
        sender.setTranslation(CGPointZero, in: view)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Add to View Hierarchy
        view.addSubview(pannableView)

        // Center Pannable View
        pannableView.center = view.center

        // Initialize Swipe Gesture Recognizer
        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(didPan(_:)))

        // Add Swipe Gesture Recognizer
        pannableView.addGestureRecognizer(panGestureRecognizer)
    }
}
```

xu-baolin replied:
> 
> ```
> 5: Offset(-100, 0),
> 6: Offset(100, 0),
> 7: Offset(0, -100),
> 8: Offset(0, 100),
> ```
> 
> what's the behavior on native for cases like this?

The draggable object will remain stationary.

xu-baolin committed:
adjust pan gesture behavior

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin commented:
Ping @dkwingsmt : )

dkwingsmt commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -51,24 +51,45 @@ enum DragStartBehavior {
+  /// If the user is dragging with 5 pointers at the same time, each having
+  /// \[+10, +20, +33, -1, -12\] pixels of offset, the recognizer will report a
+  /// delta of (+33) + (-12) = 21 pixels.
+  ///
```
Mention the panning algorithm here, and change the first paragraph.

And I feel like the name `maxAllPointers` doesn't apply to panning, but I can't think of a better name either. 

Maybe `marginPointers`, in contrast to `latestPointer`, to indicate that the result comes from all the outermost ones. For one-dimensional movement, they're positive-most one and the negavie-most one. For two-dimensional movement, they're essentially all pointers. 

This is my suggestion of the entire documentation:
```dart
  /// All active pointers will be tracked, and the result is computed from the "outermost" pointers.
  ///
  /// The scrolling offset is determined by the maximum deltas of both directions.
  ///
  /// If the user is dragging with 3 pointers at the same time, each having
  /// \[+10, +20, +33\] pixels of offset, the recognizer will report a delta of 33 pixels.
  ///
  /// If the user is dragging with 5 pointers at the same time, each having
  /// \[+10, +20, +33, -1, -12\] pixels of offset, the recognizer will report a
  /// delta of (+33) + (-12) = 21 pixels.
  ///
  /// The panning offset is the average of all pointers.
  ///
  /// If the user is dragging ... [insert an example here].
  ///
  /// This is the behavior typically seen on iOS.
```

dkwingsmt replied:
We might need a native speaker for the final verdict. @Piinks May we have your idea? Does the doc and the variable name look clear enough to you?

xu-baolin replied:
I also provide two names for reference, `mergePointers` or `conbinePointers`.

dkwingsmt replied:
I suggest against the word `sum`, because it's not a summation. The rule `sumAllPointers` literally adds all deltas together, which leads to twice the scrolling delta for two pointers. Other than that, I think either `extremePointers` or `boundaryPointers` sound ok. (`directionalMaxPointers` sounds a little too lengthy for me for no extra benefits.)

dkwingsmt replied:
If we want a verb, maybe we can use `averageBoundaryPointers`, which applies to both the one-directial algorithm and the panning algorithm.

dkwingsmt reviewed (approved):
LGTM

xu-baolin committed:
rename and improve docs

xu-baolin committed:
remove unnecessary import

xu-baolin committed:
resolve conflict

xu-baolin commented:
@dkwingsmt hi, could you help to see why the Google test failed, thank you :)

dkwingsmt commented:
It seems to be some compiling error (import not found) that has nothing to do with your change. Maybe wait for a day and try again.

xu-baolin commented:
> It seems to be some compiling error (import not found) that has nothing to do with your change. Maybe wait for a day and try again.

OK, will landing after G-test restored.

dkwingsmt committed:
Merge branch 'master' into xbl240105

xu-baolin committed:
Merge branch 'master' into xbl240105
